wtxmgr/query.go

```diff
@@ -0,0 +1,462 @@
+ */
+
+package wtxmgr
+
+import (
+	"fmt"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// CreditRecord contains metadata regarding a transaction credit for a known
+// transaction.  Further details may be looked up by indexing a wire.MsgTx.TxOut
+// with the Index field.
+type CreditRecord struct {
+	Index  uint32
+	Amount btcutil.Amount
+	Spent  bool
+	Change bool
+}
+
+// DebitRecord contains metadata regarding a transaction debit for a known
+// transaction.  Further details may be looked up by indexing a wire.MsgTx.TxIn
+// with the Index field.
+type DebitRecord struct {
+	Amount btcutil.Amount
+	Index  uint32
+}
+
+// TxDetails is intended to provide callers with access to rich details
+// regarding a relevant transaction and which inputs and outputs are credit or
+// debits.
+type TxDetails struct {
+	TxRecord
+	Block   BlockMeta
+	Credits []CreditRecord
+	Debits  []DebitRecord
+}
+
+// TxDetails looks up all recorded details regarding a transaction with some
+// hash.  In case of a hash collision, the most recent transaction with a
+// matching hash is returned.
+//
+// Not finding a transaction with this hash is not an error.  In this case,
+// a nil TxDetails is returned.
+func (s *Store) TxDetails(txHash *wire.ShaHash) (*TxDetails, error) {
+	var details *TxDetails
+	err := scopedView(s.namespace, func(ns walletdb.Bucket) error {
+		var err error
+
+		// First, check whether there exists an unmined transaction with this
+		// hash.  Use it if found.
+		v := existsRawUnmined(ns, txHash[:])
+		if v != nil {
+			details, err = s.unminedTxDetails(ns, txHash, v)
+			return err
+		}
+
+		// Otherwise, if there exists a mined transaction with this matching
+		// hash, skip over to the newest and begin fetching all details.
+		k, v := latestTxRecord(ns, txHash)
+		if v == nil {
+			// not found
+			return nil
+		}
+
+		details, err = s.txDetails(ns, txHash, k, v)
+		return err
+	})
+	return details, err
+}
+
+// UniqueTxDetails looks up all recorded details for a transaction recorded
+// mined in some particular block, or an unmined transaction if block is nil.
+//
+// Not finding a transaction with this hash from this block is not an error.  In
+// this case, a nil TxDetails is returned.
+func (s *Store) UniqueTxDetails(txHash *wire.ShaHash, block *Block) (*TxDetails, error) {
+	var details *TxDetails
+	err := scopedView(s.namespace, func(ns walletdb.Bucket) error {
+		var err error
+		if block == nil {
+			v := existsRawUnmined(ns, txHash[:])
+			if v == nil {
+				return nil
+			}
+			details, err = s.unminedTxDetails(ns, txHash, v)
+			return err
+		}
+
+		k, v := existsTxRecord(ns, txHash, block)
+		if v == nil {
+			return nil
+		}
+		details, err = s.txDetails(ns, txHash, k, v)
+		return err
+	})
+	return details, err
+}
+
+func (s *Store) txDetails(ns walletdb.Bucket, txHash *wire.ShaHash, recKey, recVal []byte) (*TxDetails, error) {
```

Comment:
This would be better defined before it's called by `TxDetails` and `UniqueTxDetails`.